### DISCUSSION TITLE
Could not load debug symbols for binary (c++ compiled with gcc12)

### DISCUSSION THREAD

**Discussion by qathy** [google/orbit#4859] [Q&A] [answered]
Hello,

Despite my binary contains debug_info, Orbit complains about symbol loading.
What are the prerequisites I'm missing when building my binary to help Orbit loading the symbols?
Is the build id mandatory?
How may I include it in my binary?

`file ut_sec`
`ut_sec: ELF 64-bit LSB executable, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, with debug_info, not stripped`

`nm -C ut_sec | grep sec_fed_close_ctx`
`0000000000431568 T sec_fed_close_ctx`

Orbit details:

 Could not load debug symbols for "/home/lucp/libsec/lib/ut_sec": Failed to find symbols for module "/home/lucp/libsec/lib/ut_sec" with build_id="":
- Unable to find local symbols for module "/home/lucp/libsec/lib/ut_sec": build id is empty.
- Not able to search for symbols on the instance.

Also: Could not load fallback symbols for "/home/lucp/libsec/lib/ut_sec":
  * Could not find module file itself in cache: Unable to find object file in cache for module "/home/lucp/libsec/lib/ut_sec"
  * Could not find module file itself locally.
  * Could not search for module file itself on the instance.

**Comment by beckerhe** [comment]
Hi qathy,
can you post the contents of the latest log file? You should be able to find it in `$HOME/.orbitprofiler/logs`.

> Is the build id mandatory?

It is for some scenarios - like debuginfo stores (i.e. /usr/lib/debug). I'm not sure why you should get this error messages if the debug info is in the same binary. But the logfile should give us all the information to figure out what's going on.

> How may I include it in my binary?

Most linkers (that is at least GNU ld, lld, gold) have an option `--build-id`. I thought it swichted on by default in all of those, but I might be mistaken. With `readelf -SW /path/to/binary | grep .note.gnu.build-id` you can also if a build is present in your program.

**Reply by qathy** [reply]
Thank you for your answer.

I had a look into the logs but it doesn't help me understanding what's going on.
Hope you'll find something interesting in the attached file.

[Orbit-2023_12_12_13_56_21-5911.log](https://github.com/google/orbit/files/13658816/Orbit-2023_12_12_13_56_21-5911.log)

**Reply by beckerhe** [reply]
That's indeed odd. I don't see it trying to load symbols for the main executable.

Or you running the process that you wanna profile in some kind of sandbox environment - like a Docker container?

Can you also post the memory map of your process (`cat /proc/$PID/maps`) I'm wondering if there is something weird going on their.

**Reply by qathy** [reply]
Hello,
You'll find this map below.
By the way, I worked around the issue re-building the binary on the same Ubuntu 22.04 system where Orbit is installed. 
Hence, using g++ 11.
The original binary was built on an Ubuntu 20.04 using g++ 12 configured without build-id enabled.

00400000-0040f000 r--p 00000000 fd:00 9571118                            /home/xxx/libsec/lib/ut_sec
0040f000-005b6000 r-xp 0000f000 fd:00 9571118                            /home/xxx/libsec/lib/ut_sec
005b6000-0069b000 r--p 001b6000 fd:00 9571118                            /home/xxx/libsec/lib/ut_sec
0069b000-0069d000 r--p 0029a000 fd:00 9571118                            /home/xxx/libsec/lib/ut_sec
0069d000-006a0000 rw-p 0029c000 fd:00 9571118                            /home/xxx/libsec/lib/ut_sec
006a0000-006a4000 rw-p 00000000 00:00 0
016da000-021b8000 rw-p 00000000 00:00 0                                  [heap]
7f5ca2600000-7f5ca2628000 r--p 00000000 fd:00 6817878                    /usr/lib/x86_64-linux-gnu/libc.so.6
7f5ca2628000-7f5ca27bd000 r-xp 00028000 fd:00 6817878                    /usr/lib/x86_64-linux-gnu/libc.so.6
7f5ca27bd000-7f5ca2815000 r--p 001bd000 fd:00 6817878                    /usr/lib/x86_64-linux-gnu/libc.so.6
7f5ca2815000-7f5ca2819000 r--p 00214000 fd:00 6817878                    /usr/lib/x86_64-linux-gnu/libc.so.6
7f5ca2819000-7f5ca281b000 rw-p 00218000 fd:00 6817878                    /usr/lib/x86_64-linux-gnu/libc.so.6
7f5ca281b000-7f5ca2828000 rw-p 00000000 00:00 0
7f5ca2a00000-7f5ca2a9a000 r--p 00000000 fd:00 6833249                    /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.30
7f5ca2a9a000-7f5ca2bab000 r-xp 0009a000 fd:00 6833249                    /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.30
7f5ca2bab000-7f5ca2c1a000 r--p 001ab000 fd:00 6833249                    /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.30
7f5ca2c1a000-7f5ca2c1b000 ---p 0021a000 fd:00 6833249                    /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.30
7f5ca2c1b000-7f5ca2c26000 r--p 0021a000 fd:00 6833249                    /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.30
7f5ca2c26000-7f5ca2c29000 rw-p 00225000 fd:00 6833249                    /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.30
7f5ca2c29000-7f5ca2c2c000 rw-p 00000000 00:00 0
7f5ca2e00000-7f5ca2e92000 r-xp 00000000 fd:00 7617387                    /usr/netlocal/curl/curl-8.4.0-openssl-3.0.11/lib/libcurl.so.4
7f5ca2e92000-7f5ca3092000 ---p 00092000 fd:00 7617387                    /usr/netlocal/curl/curl-8.4.0-openssl-3.0.11/lib/libcurl.so.4
7f5ca3092000-7f5ca3094000 r--p 00092000 fd:00 7617387                    /usr/netlocal/curl/curl-8.4.0-openssl-3.0.11/lib/libcurl.so.4
7f5ca3094000-7f5ca3098000 rw-p 00094000 fd:00 7617387                    /usr/netlocal/curl/curl-8.4.0-openssl-3.0.11/lib/libcurl.so.4
7f5ca3200000-7f5ca360c000 r-xp 00000000 fd:00 7610872                    /usr/netlocal/openssl/openssl-3.0.11/lib/libcrypto.so.3
7f5ca360c000-7f5ca380b000 ---p 0040c000 fd:00 7610872                    /usr/netlocal/openssl/openssl-3.0.11/lib/libcrypto.so.3
7f5ca380b000-7f5ca3866000 r--p 0040b000 fd:00 7610872                    /usr/netlocal/openssl/openssl-3.0.11/lib/libcrypto.so.3
7f5ca3866000-7f5ca3869000 rw-p 00466000 fd:00 7610872                    /usr/netlocal/openssl/openssl-3.0.11/lib/libcrypto.so.3
7f5ca3869000-7f5ca386c000 rw-p 00000000 00:00 0
7f5ca3a00000-7f5ca3a9b000 r-xp 00000000 fd:00 7610873                    /usr/netlocal/openssl/openssl-3.0.11/lib/libssl.so.3
7f5ca3a9b000-7f5ca3c9a000 ---p 0009b000 fd:00 7610873                    /usr/netlocal/openssl/openssl-3.0.11/lib/libssl.so.3
7f5ca3c9a000-7f5ca3ca3000 r--p 0009a000 fd:00 7610873                    /usr/netlocal/openssl/openssl-3.0.11/lib/libssl.so.3
7f5ca3ca3000-7f5ca3ca8000 rw-p 000a3000 fd:00 7610873                    /usr/netlocal/openssl/openssl-3.0.11/lib/libssl.so.3
7f5ca3d10000-7f5ca3d15000 rw-p 00000000 00:00 0
7f5ca3d15000-7f5ca3d16000 r--p 00000000 fd:00 6817886                    /usr/lib/x86_64-linux-gnu/libdl.so.2
7f5ca3d16000-7f5ca3d17000 r-xp 00001000 fd:00 6817886                    /usr/lib/x86_64-linux-gnu/libdl.so.2
7f5ca3d17000-7f5ca3d18000 r--p 00002000 fd:00 6817886                    /usr/lib/x86_64-linux-gnu/libdl.so.2
7f5ca3d18000-7f5ca3d19000 r--p 00002000 fd:00 6817886                    /usr/lib/x86_64-linux-gnu/libdl.so.2
7f5ca3d19000-7f5ca3d1a000 rw-p 00003000 fd:00 6817886                    /usr/lib/x86_64-linux-gnu/libdl.so.2
7f5ca3d1a000-7f5ca3d1d000 r--p 00000000 fd:00 6832475                    /usr/lib/x86_64-linux-gnu/libgcc_s.so.1
7f5ca3d1d000-7f5ca3d34000 r-xp 00003000 fd:00 6832475                    /usr/lib/x86_64-linux-gnu/libgcc_s.so.1
7f5ca3d34000-7f5ca3d38000 r--p 0001a000 fd:00 6832475                    /usr/lib/x86_64-linux-gnu/libgcc_s.so.1
7f5ca3d38000-7f5ca3d39000 r--p 0001d000 fd:00 6832475                    /usr/lib/x86_64-linux-gnu/libgcc_s.so.1
7f5ca3d39000-7f5ca3d3a000 rw-p 0001e000 fd:00 6832475                    /usr/lib/x86_64-linux-gnu/libgcc_s.so.1
7f5ca3d3a000-7f5ca3d48000 r--p 00000000 fd:00 6817887                    /usr/lib/x86_64-linux-gnu/libm.so.6
7f5ca3d48000-7f5ca3dc4000 r-xp 0000e000 fd:00 6817887                    /usr/lib/x86_64-linux-gnu/libm.so.6
7f5ca3dc4000-7f5ca3e1f000 r--p 0008a000 fd:00 6817887                    /usr/lib/x86_64-linux-gnu/libm.so.6
7f5ca3e1f000-7f5ca3e20000 r--p 000e4000 fd:00 6817887                    /usr/lib/x86_64-linux-gnu/libm.so.6
7f5ca3e20000-7f5ca3e21000 rw-p 000e5000 fd:00 6817887                    /usr/lib/x86_64-linux-gnu/libm.so.6
7f5ca3e21000-7f5ca3e23000 rw-p 00000000 00:00 0
7f5ca3e23000-7f5ca3e24000 r--p 00000000 fd:00 6817908                    /usr/lib/x86_64-linux-gnu/libpthread.so.0
7f5ca3e24000-7f5ca3e25000 r-xp 00001000 fd:00 6817908                    /usr/lib/x86_64-linux-gnu/libpthread.so.0
7f5ca3e25000-7f5ca3e26000 r--p 00002000 fd:00 6817908                    /usr/lib/x86_64-linux-gnu/libpthread.so.0
7f5ca3e26000-7f5ca3e27000 r--p 00002000 fd:00 6817908                    /usr/lib/x86_64-linux-gnu/libpthread.so.0
7f5ca3e27000-7f5ca3e28000 rw-p 00003000 fd:00 6817908                    /usr/lib/x86_64-linux-gnu/libpthread.so.0
7f5ca3e36000-7f5ca3e38000 rw-p 00000000 00:00 0
7f5ca3e38000-7f5ca3e3a000 r--p 00000000 fd:00 6817870                    /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
7f5ca3e3a000-7f5ca3e64000 r-xp 00002000 fd:00 6817870                    /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
7f5ca3e64000-7f5ca3e6f000 r--p 0002c000 fd:00 6817870                    /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
7f5ca3e70000-7f5ca3e72000 r--p 00037000 fd:00 6817870                    /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
7f5ca3e72000-7f5ca3e74000 rw-p 00039000 fd:00 6817870                    /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2
7ffec18ee000-7ffec190f000 rw-p 00000000 00:00 0                          [stack]
7ffec19a7000-7ffec19ab000 r--p 00000000 00:00 0                          [vvar]
7ffec19ab000-7ffec19ad000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 --xp 00000000 00:00 0                  [vsyscall]

**Reply by beckerhe** [reply, marked as answer]
Hmm, interesting. I still have no clue why Orbit is not even trying to load symbols for ut_sec (at least according to the log file).

But anyway. I'm glad that it works now. :-)
